syntax = "proto3";

service PaymentGateway {
    rpc AuthenticateClient (AuthRequest) returns (AuthResponse);
    rpc CheckBalance (BalanceRequest) returns (BalanceResponse);
    rpc ProcessPayment (PaymentRequest) returns (PaymentResponse);
    rpc ViewTransactionHistory (TransactionHistoryRequest) returns (TransactionHistoryResponse);
    rpc LogoutClient (LogoutRequest) returns (LogoutResponse);
}

message AuthRequest {
    string username = 1;
    string password = 2;
}

message AuthResponse {
  bool authenticated = 1;
  string token = 2;
  string message = 3;  // Add this field for detailed responses
}

message BalanceRequest {}  // ✅ No token field inside message
message BalanceResponse {
    double balance = 1;
}

message PaymentRequest {
    string to_account = 1;
    double amount = 2;
    string transaction_id=3;
    
}

message PaymentResponse {
    bool success = 1;
    string message = 2;
    string transaction_id = 3; // adding this field
}

message LogoutRequest {}

message LogoutResponse {
    bool success = 1;
    string message = 2;
}


// Define response for transaction history
message Transaction {
    string transaction_id = 1;
    string from_account = 2;  // ✅ Rename "from" to "from_account"
    string to_account = 3;
    double amount = 4;
    string timestamp = 5;
    string status = 6;
}


message TransactionHistoryRequest {}

message TransactionHistoryResponse {
    repeated Transaction transactions = 1;
}
